src/sql/workbench/parts/notebook/common/models/notebookInput.ts

```diff
@@ -104,22 +122,42 @@ export class NotebookEditorModel extends EditorModel {
 			// Request serialization so trusted state is preserved but don't update the model
 			this.sendNotebookSerializationStateChange();
 		} else {
-			// For all other changes, update the backing model with the latest contents
 			let notebookModel = this.getNotebookModel();
+			let editAppliedSuccessfully = false;
 			if (notebookModel && this.textEditorModel && this.textEditorModel.textEditorModel) {
-				let content = JSON.stringify(notebookModel.toJSON(), undefined, '    ');
-				let model = this.textEditorModel.textEditorModel;
-				let endLine = model.getLineCount();
-				let endCol = model.getLineMaxColumn(endLine);
-
-				this.textEditorModel.textEditorModel.applyEdits([{
-					range: new Range(1, 1, endLine, endCol),
-					text: content
-				}]);
+				if (contentChange && contentChange.cells && contentChange.cells[0]) {
+					if (type === NotebookChangeType.CellSourceUpdated) {
+						if (this.notebookTextFileModel.transformAndApplyEditForSourceUpdate(contentChange, this.textEditorModel)) {
+							editAppliedSuccessfully = true;
+						}
+					} else if (type === NotebookChangeType.CellOutputUpdated) {
+						if (this.notebookTextFileModel.transformAndApplyEditForOutputUpdate(contentChange, this.textEditorModel)) {
+							editAppliedSuccessfully = true;
+						}
+					} else if (type === NotebookChangeType.CellOutputCleared) {
+						if (this.notebookTextFileModel.transformAndApplyEditForClearOutput(contentChange, this.textEditorModel)) {
+							editAppliedSuccessfully = true;
+						}
+					} else if (type === NotebookChangeType.CellExecuted) {
+						if (this.notebookTextFileModel.transformAndApplyEditForCellUpdated(contentChange, this.textEditorModel)) {
+							editAppliedSuccessfully = true;
+						}
+					}
+				}
+				// If edit was already applied, skip replacing entire text model
+				if (editAppliedSuccessfully) {
+					return;
+				}
+				this.replaceEntireTextEditorModel(notebookModel, type);
+				this._lastEditFullReplacement = true;
 			}
 		}
 	}
 
+	public replaceEntireTextEditorModel(notebookModel: INotebookModel, type: NotebookChangeType) {
+		this.notebookTextFileModel.replaceEntireTextEditorModel(notebookModel, type, this.textEditorModel);
```

Comment:
Seems a little silly to have a one line method like this

Reply:
Yeah this was just so that the tests could access this method. I can remove if you think it's silly and try to work around it.